Fix some ClassLoader leaks from the BuildMojo

stuartwdouglas opened the pull request:
Hopefully this will fix some of the CI issues

stuartwdouglas committed:
Fix some ClassLoader leaks from the BuildMojo

Hopefully this will fix some of the CI issues

geoand commented on devtools/maven/src/main/java/io/quarkus/maven/BuildMojo.java:
```diff
@@ -133,9 +133,10 @@ public BuildMojo() {
         MojoLogger.logSupplier = this::getLog;
     }
 
+    boolean pause = true;
```
This doesn't seem to do be used anywhere

geoand commented on devtools/maven/src/main/java/io/quarkus/maven/BuildMojo.java:
```diff
@@ -201,6 +202,7 @@ public void execute() throws MojoExecutionException {
             QuarkusBootstrap.Builder builder = QuarkusBootstrap.builder()
                     .setAppArtifact(appArtifact)
                     .setMavenArtifactResolver(resolver)
+                    .setIsolateDeployment(true)
```
Is there any chance of this causing unexpected behavior?

stuartwdouglas replied:
I think the current failures are because close is being called twice, which means it attempts to define the driver remover twice

geoand replied:
OK, let's see what CI says

aloubyansky commented on independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/DriverRemover.java:
```diff
@@ -0,0 +1,25 @@
+import java.util.Enumeration;
+import org.jboss.logging.Logger;
+
+public class DriverRemover implements Runnable {
```
There are two of these classes. Is it supposed to be one for the runtime and one for the deployment CL?

stuartwdouglas replied:
oops, must have copied instead of moved

gsmet commented:
The MySQL native test is failing with:

```
2020-08-27T09:22:59.8943464Z com.oracle.svm.core.util.UserError$UserException: com.oracle.svm.hosted.substitute.DeletedElementException: Unsupported method com.mysql.cj.jdbc.AbandonedConnectionCleanupThread.uncheckedShutdown() is reachable: The declaring class of this element has been substituted, but this element is not present in the substitution class
2020-08-27T09:22:59.8944140Z To diagnose the issue, you can add the option --report-unsupported-elements-at-runtime. The unsupported element is then reported at run time when it is accessed the first time.
2020-08-27T09:22:59.8944304Z Detailed message:
2020-08-27T09:22:59.8944420Z Trace: 
2020-08-27T09:22:59.8944549Z 	at parsing io.quarkus.jdbc.mysql.runtime.MySQLRecorder$1.run(MySQLRecorder.java:21)
2020-08-27T09:22:59.8944698Z Call path from entry point to io.quarkus.jdbc.mysql.runtime.MySQLRecorder$1.run(): 
2020-08-27T09:22:59.8944850Z 	at io.quarkus.jdbc.mysql.runtime.MySQLRecorder$1.run(MySQLRecorder.java:21)
2020-08-27T09:22:59.8944995Z 	at sun.nio.ch.EPollPort$EventHandlerTask.run(EPollPort.java:300)
2020-08-27T09:22:59.8945128Z 	at java.lang.Thread.run(Thread.java:834)
2020-08-27T09:22:59.8945240Z 	at com.oracle.svm.core.thread.JavaThreads.threadStartRoutine(JavaThreads.java:517)
2020-08-27T09:22:59.8945560Z 	at com.oracle.svm.core.posix.thread.PosixJavaThreads.pthreadStartRoutine(PosixJavaThreads.java:193)
2020-08-27T09:22:59.8945741Z 	at com.oracle.svm.core.code.IsolateEnterStub.PosixJavaThreads_pthreadStartRoutine_e1f4a8c0039f8337338252cd8734f63a79b5e3df(generated:0)
2020-08-27T09:22:59.8945853Z 
2020-08-27T09:22:59.8945983Z 	at com.oracle.svm.core.util.UserError.abort(UserError.java:79)
2020-08-27T09:22:59.8946130Z 	at com.oracle.svm.hosted.FallbackFeature.reportAsFallback(FallbackFeature.java:217)
2020-08-27T09:22:59.8946265Z 	at com.oracle.svm.hosted.NativeImageGenerator.runPointsToAnalysis(NativeImageGenerator.java:753)
2020-08-27T09:22:59.8946430Z 	at com.oracle.svm.hosted.NativeImageGenerator.doRun(NativeImageGenerator.java:538)
2020-08-27T09:22:59.8946585Z 	at com.oracle.svm.hosted.NativeImageGenerator.lambda$run$0(NativeImageGenerator.java:451)
2020-08-27T09:22:59.8946746Z 	at java.base/java.util.concurrent.ForkJoinTask$AdaptedRunnableAction.exec(ForkJoinTask.java:1407)
2020-08-27T09:22:59.8946902Z 	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
2020-08-27T09:22:59.8969292Z 	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
2020-08-27T09:22:59.8970065Z 	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
2020-08-27T09:22:59.8970249Z 	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
2020-08-27T09:22:59.8970401Z 	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:177)
2020-08-27T09:22:59.8970802Z Caused by: com.oracle.graal.pointsto.constraints.UnsupportedFeatureException: com.oracle.svm.hosted.substitute.DeletedElementException: Unsupported method
```

gsmet commented:
Apparently, it didn't fix the JVM issues.

stuartwdouglas commented:
Turns out these don't help without this change: https://github.com/smallrye/smallrye-config/pull/383/files

My local testing included it, but I did not realise how important it was, as I was thinking that subsequent runs would clear the ThreadLocal (which then don't, because different ClassLoader so new ThreadLocal).

stuartwdouglas commented:
@kenfinnigan @dmlloyd who takes care of config releases? Any chance we could get a release with this fixed in the near future?

stuartwdouglas commented:
Converting to draft, as if it were merged in its current state it would actually make the situation worse with the CL isolation.

dmlloyd commented:
/cc @radcortez

kenfinnigan commented:
Roberto is PTO for the rest of the week, but anyone should be able to trigger a release

kenfinnigan commented:
Anyone know if we're ok to release SRye Config from master, or the fix needs to be backported to only include this change in 1.8.x?

Currently master is aligned with 1.9.0, so not sure if there are lots of new features we don't necessarily need/want right now

kenfinnigan commented:
1.8.6 of SRye Config has been released.

Might be half an hour or so before it's available in central for use by CI

stuartwdouglas committed:
Limit maven metaspace during the build

This ensures it will clean up old ClassLoaders,
instead of just allocating more memory.

stuartwdouglas committed:
Fix ClassLoader leak in the docs build

Also raise the metaspace limit, there is a bit of
trial and error here to find out what works.

geoand commented:
Can you please rebase this onto master to pick up the latest CI changes?

stuartwdouglas commented:
The JDK 14 job actually passed last run, but the mutiny test caused the other 3 to fail. I have rebased, so lets hope it is all green.

geoand commented:
Cool, thanks

jaikiran commented:
Almost there. The failures (in both these failing jobs) is:

```
2020-08-28T08:58:32.5582106Z [ERROR] Tests run: 6, Failures: 0, Errors: 2, Skipped: 0, Time elapsed: 5.318 s <<< FAILURE! - in io.quarkus.it.mongodb.panache.reactive.ReactiveMongodbPanacheResourceTest
2020-08-28T08:58:32.5725011Z [ERROR] io.quarkus.it.mongodb.panache.reactive.ReactiveMongodbPanacheResourceTest.testReactiveBookEntity  Time elapsed: 1.877 s  <<< ERROR!
2020-08-28T08:58:32.5737782Z org.awaitility.core.ConditionTimeoutException: Condition with lambda expression in io.quarkus.it.mongodb.panache.reactive.ReactiveMongodbPanacheResourceTest that uses io.quarkus.it.mongodb.panache.reactive.ReactiveMongodbPanacheResourceTest$IntegerAdder was not fulfilled within 1 seconds.
2020-08-28T08:58:32.5913682Z 	at org.awaitility.core.ConditionAwaiter.await(ConditionAwaiter.java:165)
2020-08-28T08:58:32.5914946Z 	at org.awaitility.core.CallableCondition.await(CallableCondition.java:78)
2020-08-28T08:58:32.5953090Z 	at org.awaitility.core.CallableCondition.await(CallableCondition.java:26)
2020-08-28T08:58:32.5954780Z 	at org.awaitility.core.ConditionFactory.until(ConditionFactory.java:895)
2020-08-28T08:58:32.6093463Z 	at org.awaitility.core.ConditionFactory.until(ConditionFactory.java:864)
2020-08-28T08:58:32.6094706Z 	at io.quarkus.it.mongodb.panache.reactive.ReactiveMongodbPanacheResourceTest.callReactiveBookEndpoint(ReactiveMongodbPanacheResourceTest.java:208)
```
I've a PR open for that one https://github.com/quarkusio/quarkus/pull/11690

geoand commented:
I'm going to merge this as it does seem to further address the memory concerns
